File: 프로그래머스/2/148653. 마법의 엘리베이터/마법의 엘리베이터.py
from collections import defaultdict
def solution(storey):
    answer = 0
    
    visited = defaultdict(int)
    cur = [storey]
    while cur:
        temp = []
        for c in cur:
            for d in range(10):
                if c % 10**d != 0:
                    interest = c - 10**(d-1)
                    if visited[interest] == 0:
                        visited[interest] = 1
                        temp.append(interest)
                    
                    break
            for d in range(10):
                if c % 10**d != 0:
                    interest = c + 10 ** (d-1)
                    if visited[interest] == 0:
                        visited[interest] = 1
                        temp.append(interest)
                    break
        cur = temp
        answer += 1
        if visited[0] == 1:
            break
                                    
    return answer

File: 프로그래머스/2/148653. 마법의 엘리베이터/마법의 엘리베이터.java
import java.util.*;

class Solution {
    int answer;
    int storey;
    
    public int solution(int storey) {
        init(storey);
        solve();
        return answer;
    }
    
    void init(int storey) {
        this.storey = storey;
        this.answer = 0;
    }
    
    void solve() {
        while (storey > 0) {
            int digit = storey % 10;
            storey = storey / 10;
            if (digit == 5) {
                if (storey % 10 >= 5) {
                    answer = answer + (10 - digit);
                    storey++;
                }
                else {
                    answer = answer + digit;    
                }
            }
            else if (digit > 5) {
                answer = answer + (10 - digit);
                storey++;
            }
            else {
                answer = answer + digit;
            }

        }
    }
}